### PR TITLE
New functionality that can add the localhost IP address as a valid igress for DockerHosts

### PR DESCRIPTION
**Pull request by Graeme-Miller**

### CODE REVIEW

**Comment by Graeme-Miller** [comment]
Should fix:
https://github.com/brooklyncentral/clocker/issues/134
https://github.com/brooklyncentral/clocker/issues/215

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java [hunk 1]
```diff
@@ -591,26 +594,31 @@ public void addIpPermissions(Collection<IpPermission> permissions) {
...
-                .toPort(sensors().get(DockerHost.DOCKER_SSL_PORT))
-                .cidrBlock(localhost)
-                .build();
+        List<IpPermission> permissions = MutableList.of();
+
+        String publicIpCIDR = LocalhostExternalIpLoader.getLocalhostIpWithin(Duration.minutes(1)) + "/32";
+        permissions.addAll(getClockerPermisionsForCIDR(publicIpCIDR));
+
+        Boolean addLocalhost = config().get(ADD_LOCALHOST_PERMISSION);
+        if(Boolean.TRUE.equals(addLocalhost)) {
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]:
```Java
if (config().get(ADD_LOCALHOST_PERMISSION)) {
```

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java [hunk 2]
```diff
@@ -591,26 +594,31 @@ public void addIpPermissions(Collection<IpPermission> permissions) {
...
+            String localhostAddress = null;
+            try {
+                localhostAddress = InetAddress.getLocalHost().getHostAddress();
+            } catch (UnknownHostException e) {
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]: Don't need `try`/`catch` block

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java (hunk 1 above)
Comment (sjcorbett) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]: Boolean.TRUE.equals is uglier but prevents null pointer exceptions when someone nulls the config for whatever reason.

Comment (sjcorbett) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]: Alternatively if you are using `0.9.0` now you could define the config key with a `notNull` constraint and my comment above is irrelevant.

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java [hunk 3]
```diff
@@ -591,26 +594,31 @@ public void addIpPermissions(Collection<IpPermission> permissions) {
...
+            }
+
+            String localhostCIDR = localhostAddress + "/32";
+            if (localhostAddress != null && !localhostAddress.isEmpty() && !publicIpCIDR.equals(localhostCIDR)) {
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]:
```Java
if (Strings.isNotEmpty(localhostAddress) && !publicIpCIDR.equals(localhostCIDR)) {
```

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java [hunk 4]
```diff
@@ -622,9 +630,29 @@ public void addIpPermissions(Collection<IpPermission> permissions) {
             permissions.add(dockerRegistryPort);
         }
 
+        return permissions;
+    }
+
+    private List<IpPermission> getClockerPermisionsForCIDR(String CIDR) {
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]: No UPPERCASE arguments. Use `cidr`

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java (hunk 1 above)
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java]: We'll go through and add constraints when we do the upgrade, then. Change to the sort-of unsafe version, and comment the config key appropriately.

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHost.java [hunk 5]
```diff
@@ -139,6 +139,12 @@
     AttributeSensor<SubnetTier> DOCKER_HOST_SUBNET_TIER = Sensors.newSensor(SubnetTier.class,
             "docker.subnetTier", "The SubnetTier for Docker port mapping");
 
+    @SetFromFlag("addLocalhostPermission")
+    ConfigKey<Boolean> ADD_LOCALHOST_PERMISSION = ConfigKeys.newBooleanConfigKey(
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHost.java]:
```Java
// TODO add notNull constraint
```

File: docker/src/main/java/brooklyn/entity/container/docker/DockerHost.java [hunk 6]
```diff
@@ -139,6 +139,12 @@
...
+            "docker.add.localhost.permission",
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerHost.java]: `docker.addLocalhostPermission`

**Comment by grkvlt** [comment]
Thanks @Graeme-Miller have merged :frog:

**Pull request merged**